modules/core/shared/src/main/scala/io/circe/JsonNumber.scala

```diff
@@ -208,21 +215,23 @@ private[circe] final case class JsonDouble(value: Double) extends JsonNumber {
  * Represent a valid JSON number as a [[scala.Float]].
  */
 private[circe] final case class JsonFloat(value: Float) extends JsonNumber {
-  private[this] def toJavaBigDecimal = new java.math.BigDecimal(java.lang.Float.toString(value))
-
   private[circe] def toBiggerDecimal: BiggerDecimal = BiggerDecimal.fromFloat(value)
+  private[this] def toJavaBigDecimal = new JavaBigDecimal(java.lang.Float.toString(value))
+
   final def toBigDecimal: Option[BigDecimal] = Some(toJavaBigDecimal)
-  final def toBigInt: Option[BigInt] = toBigDecimal.flatMap { d =>
-    val bigInt = d.toBigInt
-    if (bigInt.floatValue() == value) Some(bigInt) else None
+  final def toBigInt: Option[BigInt] = {
+    val asBigDecimal = toJavaBigDecimal
+
+    if (JsonNumber.bigDecimalIsWhole(asBigDecimal)) Some(new BigInt(asBigDecimal.toBigInteger)) else None
   }
 
-  // Don't use value.toFloat due to floating point errors.
-  final def toDouble: Double = toJavaBigDecimal.doubleValue()
+  // Don't use `value.toFloat` due to floating point errors.
```

Comment:
:+1:, thanks for noticing. I want to add a class abstracting over `JsonFloat` and `JsonDouble` but MiMa didn't like it, so I was planning to do it separately, and can take care of this then.